### ISSUE TITLE
TypeLoader#load_project failing when a T::Struct has a nested class

### ISSUE THREAD

**Issue by jeffcarbs** [AaronC81/parlour#78]
**Describe the bug**
`Parlour::TypeLoader.load_project` fails when a T::Struct has a nested class defined outside of the original class body.

**To Reproduce**
```ruby
namespace = Parlour::TypeLoader.load_source(<<-RUBY)
  class A < T::Struct; end
  class A::B; end
RUBY

# Called within `load_project`
Parlour::ConflictResolver.new.resolve_conflicts(namespace) do |n, o|
  raise "conflict of #{o.length} objects: #{n}"
end
```

**Expected behavior**
No conflict

**Actual behavior**
Conflict

**Additional information**
I think what's happening is that when the loader encounters `A::B` it creates a top-level `A` child but without a subclass so the conflict resolution thinks they're different. This doesn't happen when the original `A` is a subclass of some other class which makes me think this is specific to T::Struct. Some other things I tried:

Using the "expanded" nesting still doesn't work.
```ruby
namespace = Parlour::TypeLoader.load_source(<<-RUBY)
  class A < T::Struct; end

  class A
    class B; end
  end
RUBY
```

Explicitly subclass from T::Struct again works
```ruby
namespace = Parlour::TypeLoader.load_source(<<-RUBY)
  class A < T::Struct; end

  class A < T::Struct
    class B; end
  end
RUBY
```

Non-T::Struct works:
```ruby
namespace = Parlour::TypeLoader.load_source(<<-RUBY)
  class A < Foo; end
  class A::B; end
RUBY
```

**Comment by AaronC81** [comment]
I think this is probably a special case that I forgot to extend to including `StructClassNamespace`, so this shouldn't be too hard to fix. Thanks!

**Comment by AaronC81** [comment]
I think I might've fixed this - please could you test the `fix-78` branch on the codebase that was causing this issue? Ta!

**Comment by jeffcarbs** [comment]
Hey @AaronC81 - sorry for the delay! I tried the `fix-78` branch out and while it does fix the repro case I provided it actually doesn't fix the "expanded" case I mentioned in the description. I looked at the test you added and here's what I think you'd want to have pass:

```diff
 it 'allows specialized classes (structs/enums) to have namespace children' do
   x = Parlour::TypeLoader.load_source(<<-RUBY)
     class A < T::Struct; end
     class A::B; end
+    class A
+      class Z; end
+    end

     class C < T::Enum; end
     class C::D; end
+    class C
+      class Y; end
+    end
   RUBY
 
   expect(x.children.length).to be 4
 
   subject.resolve_conflicts(x) { |*x| raise 'unable to resolve automatically' }
 
   expect(x.children.length).to be 2
 end
```

**Comment by jeffcarbs** [comment]
Playing around with some different permutations, I think you might want to update the test to be something like this:
```ruby
x = Parlour::TypeLoader.load_source(<<-RUBY)
  class A < T::Struct
    prop :x, String
  end
  class A::B; end
  class A
    class C; end
  end
  class A < T::Struct; end

  class Z < T::Enum
    enums do
      North = new
    end
  end
  class Z::Y; end
  class Z
    class X; end
  end
  class Z < T::Enum; end
RUBY
```

For each specialized type (struct and enum) it would test:
- 4 different colliding classes
- 2 specialized namespaces, 1 namespace, 1 class namespace
- 1 specialized namespace _with_ props/enums, 1 with empty props/enums

**Comment by AaronC81** [comment]
Thank you for that expanded test case @jeffcarbs - I think this should be good now. Please give it a try and let me know!

**Comment by jeffcarbs** [comment]
@AaronC81 - Still didn't work on my codebase and realized the "exhaustive" test cases I suggested added too much 🙂 Here's the minimal repro test case:
```ruby
class A < T::Struct
end
class A; end
```

In this test case, there's no pure `Namespace` component which means the `exactly_namespace.length == 1` piece fails. The first instance of `A` is a `StructClassNamespace` and the second is a `ClassNamespace`.

I tried commenting that line out and it fixed the issue here but broke this test: https://github.com/AaronC81/parlour/blob/master/spec/conflict_resolver_spec.rb#L186-L200. This means that what I'm suggesting in a minimal repro case probably _shouldn't_ be valid. In this case, I'm not exactly sure what ought to be allowed or not. It's valid ruby to reopen classes and add more things to it. It's not valid ruby to have a module and then make it a class. That being said, it's also impossible to know if the `A` in `class A::B` should be a module or a class without prior knowledge.

Up to you how you'd like to handle it. I sort of think that what I have above _should_ work, which would mean we'd need to update that other "does not merge enums and classes" test would need to be updated.
